core/src/test/java/kafka/test/api/CustomQuotaCallbackTest.java

```diff
@@ -0,0 +1,137 @@
+        @ClusterConfigProperty(id = 3001, key = QuotaConfig.CLIENT_QUOTA_CALLBACK_CLASS_CONFIG, value = "kafka.test.api.CustomQuotaCallbackTest$CustomQuotaCallback"),
+        @ClusterConfigProperty(id = 3002, key = QuotaConfig.CLIENT_QUOTA_CALLBACK_CLASS_CONFIG, value = "kafka.test.api.CustomQuotaCallbackTest$CustomQuotaCallback"),
+    }
+)
+@ExtendWith(ClusterTestExtensions.class)
+public class CustomQuotaCallbackTest {
+
+    private final ClusterInstance cluster;
+
+    public CustomQuotaCallbackTest(ClusterInstance clusterInstance) {
+        this.cluster = clusterInstance;
+    }
+
+    @ClusterTest
+    public void testCustomQuotaCallbackWithControllerServer() throws InterruptedException {
+        
+        try (Admin admin = cluster.admin(Map.of())) {
+            admin.createTopics(List.of(new NewTopic("topic", 1, (short) 1)));
+            TestUtils.waitForCondition(
+                () -> CustomQuotaCallback.COUNTERS.size() == 3 
+                        && CustomQuotaCallback.COUNTERS.values().stream().allMatch(counter -> counter.get() > 0), 
+                    "The CustomQuotaCallback not triggered in all controllers. "
+            );
+
+//            List<ClientQuotaAlteration> clientQuotaAlterations = List.of(
+//                new ClientQuotaAlteration(new org.apache.kafka.common.quota.ClientQuotaEntity(Map.of(CLIENT_ID, "testClient")), List.of(
+//                    new ClientQuotaAlteration.Op(QuotaConfig.PRODUCER_BYTE_RATE_OVERRIDE_CONFIG, 8000.0),
+//                    new ClientQuotaAlteration.Op(QuotaConfig.CONSUMER_BYTE_RATE_OVERRIDE_CONFIG, 8000.0))
+//                )
+//            );
+            // Reset the counters, and we expect the callback to be triggered again in all controllers
+            CustomQuotaCallback.COUNTERS.clear();
+//            admin.alterClientQuotas(clientQuotaAlterations, new AlterClientQuotasOptions());
+            admin.deleteTopics(List.of("topic"));
+            TestUtils.waitForCondition(
+                () -> CustomQuotaCallback.COUNTERS.size() == 3
+                        && CustomQuotaCallback.COUNTERS.values().stream().allMatch(counter -> counter.get() > 0), 
+                    "The CustomQuotaCallback not triggered in all controllers. "
+            );
+            
+        }
+    }
+
+
+    public static class CustomQuotaCallback implements ClientQuotaCallback {
+
+        public static final Map<String, AtomicInteger> COUNTERS = new ConcurrentHashMap<>();
+        private String nodeId;
+
+        @Override
+        public Map<String, String> quotaMetricTags(ClientQuotaType quotaType, KafkaPrincipal principal, String clientId) {
+            return Map.of();
+        }
+
+        @Override
+        public Double quotaLimit(ClientQuotaType quotaType, Map<String, String> metricTags) {
+            return 0.0;
```

Comment:
You have to return `Double.MAX_VALUE` ... otherwise, the ops can't get pass as they are rejected due to throttle.